File: 56.merge-intervals.py
#
# @lc app=leetcode id=56 lang=python3
#
# [56] Merge Intervals
#

# @lc code=start
from collections import deque


class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key=lambda x: x[0])
        res = deque([intervals[0]])

        for l, r in intervals[1:]:
            topr = res[-1][1]
            if l <= topr: # overlap occurs, merge
                res[-1][1] = max(r, topr)
            else:
                res.append([l, r])
        return res


# @lc code=end

File: RustSolutions/src/56.merge-intervals.rs
/*
 * @lc app=leetcode id=56 lang=rust
 *
 * [56] Merge Intervals
 */

// @lc code=start


impl Solution {
    pub fn merge(mut intervals: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        intervals.sort_unstable_by_key(|a| a[0]);

        let (mut min, mut max) = (intervals[0][0], intervals[0][1]);
        let mut ret = Vec::new();

        for item in &intervals[1..] {
            if max < item[0] {
                ret.push(vec![min, max]);
                min = item[0];
            }

            min = min.min(item[0]);
            max = max.max(item[1]);
        }

        ret.push(vec![min, max]);
        ret
    }
}
// @lc code=end

